File: NERP.Design/Models/Design.cs
using Common.DataAccess.Mongo;
using NERP.Common.Models;
using NERP.NetworkEntities.Models;
using System.Collections.Generic;

namespace NERP.Designs.Models
{
    public class Design : DbBaseEntity
    {
        public string ReferenceId { get; set; }
        public string Name { get; set; }
        public string Code { get; set; }
        public string Description { get; set; }
        public string DesignType { get; set; }
        public string AssignedTo { get; set; }
        public double Version { get; set; }
        public List<Document> Documents { get; set; }
        public string ClusterId { get; set; }
        public List<Homepass> Homepasses { get; set; }
        public List<BookingId> BookingIds { get; set; }
        public List<NetworkSetting> NetworkSettings { get; set; }
        public string Log { get; set; }
        public List<Costing> Boq { get; set; }
        public GeoNode BluePrint { get; set; }
        public string Status { get; set; }
        public string ProjectId { get; set; }
        public string ClusterCode { get; set; }
    }


    public class Costing
    {
        public string Level { get; set; }
        public string EntityType { get; set; }
        public string Code { get; set; }
        public string Description { get; set; }
        public string Units { get; set; }
        public double Quantity { get; set; }
        public double? MaterialCost { get; set; }
        public double? ServiceCost { get; set; }
        public string Currency { get; set; }
    }

    public class NetworkSetting
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string Position { get; set; }
        public NodeSettings StartNode { get; set; }
        public NodeSettings DestinationNode { get; set; }
        public NodeSettings FDPNode { get; set; }
        public Cabling Cabling { get; set; }
    }

    public class NodeSettings
    {
        public EntitySettings EntitySettings { get; set; }
        public FDPSettings FDPSettings { get; set; }
        public AccessorySettings AccessorySettings { get; set; }
        public CivilworkSettings CivilworkSettings { get; set; }
        public DuctSettings DuctSettings { get; set; }
        public EquipmentSettings EquipmentSettings { get; set; }
        public double FiberDrop { get; set; }
    }

    public class Cabling
    {
        public CableSettings CableSettings { get; set; }
        public SpliceClosureSettings SpliceClosureSettings { get; set; }
        public AccessorySettings AccessorySettings { get; set; }
        public DuctSettings DuctSettings { get; set; }
    }

    public class EntitySettings
    {
        public string EntityName { get; set; }
        public string DataSource { get; set; }
        public string ExistingDesignId { get; set; }
        public string ImportedFile { get; set; }
        public List<Chamber> Chambers { get; set; }
        public List<ODF> ODFs { get; set; }
        public List<Pole> Poles { get; set; }
        public List<SpliceClosure> SpliceClosures { get; set; }
        public List<Splitter> Splitters { get; set; }

    }

    public class FDPSettings
    {
        public string DataSource { get; set; }
        public string ExistingDesignId { get; set; }
        public string ImportedFile { get; set; }
        public List<FDP> Fdps { get; set; }
    }

    public class AccessorySettings
    {
        public string EntityName { get; set; }
        public string DataSource { get; set; }
        public List<Accessory> Accessories { get; set; }
        public List<Accessory> ExistingAccessories { get; set; }
    }

    public class CivilworkSettings
    {
        public List<Civilwork> Civilworks { get; set; }
    }

    public class DuctSettings
    {
        public List<Duct> Ducts { get; set; }
    }

    public class EquipmentSettings
    {
        public string EntityName { get; set; }
        public string DataSource { get; set; }
        public List<string> Equipments { get; set; }
    }

    public class CableSettings
    {
        public string CableType { get; set; }
        public string DataSource { get; set; }
        public List<Cable> Cables { get; set; }
        public Pole Pole { get; set; }
        public Chamber Chamber { get; set; }
        public double ChamberInterval { get; set; }
        public double CableLoop { get; set; }
        public double LoopInterval { get; set; }
        public double LoopAtClosure { get; set; }
        public double ConnectionPercent { get; set; }
    }

    public class SpliceClosureSettings
    {
        public string EntityName { get; set; }
        public string DataSource { get; set; }
        public List<SpliceClosure> SpliceClosures { get; set; }
        public bool StraightJoin { get; set; }
        public double DrumLength { get; set; }
        public bool IntersectionJoints { get; set; }
    }
}

File: NERP.RND.Web/ClientApp/src/app/designs/models/design.ts
import { DbBaseEntity } from '../../common/models/dbBaseEntity';
import { Document } from '../../common/models/document';
import { BookingId } from '../../common/models/bookingId';
import { Cable } from '../../entities/models/cable';
import { Chamber } from '../../entities/models/chamber';
import { Duct } from '../../entities/models/duct';
import { Pole } from '../../entities/models/pole';
import { SpliceClosure } from '../../entities/models/spliceClosure';
import { Splitter } from '../../entities/models/splitter';
import { Accessory } from '../../accessories/models/accessory';
import { Civilwork } from '../../civilworks/models/civilwork';

import { GeoNode } from '../../common/geo/geoNode';
import { ODF } from '../../entities/models/odf';
import { FDP } from '../../entities/models/fdp';
import { Homepass } from '../../homepasses/models/homepass';
import { PoP } from '../../entities/models/pop';

export class Design extends DbBaseEntity {
    referenceId: string;
    name: string;
    code: string;
    description: string;
    designType: string;
    assignedTo: string;
    version: number = 1;
    documents: Document[] = new Array<Document>();
    clusterId: string;
    homepasses: Homepass[]; 
    bookingIds: BookingId[] = new Array<BookingId>();
    networkSettings: NetworkSetting[] = new Array<NetworkSetting>();
    log: string = "";
    bluePrint?: GeoNode;
    boq: Costing[];

    generatedCodes: CodeId[];

}

export class NetworkSetting {
    id: string;
    name: string;
    position: number;
    startNode: NodeSettings = new NodeSettings();
    destinationNode: NodeSettings = new NodeSettings();
    fdpNode?: NodeSettings = new NodeSettings();
    cabling: Cabling = new Cabling();
}

export class NodeSettings {
    entitySettings: EntitySettings = new EntitySettings();
    fdpSettings?: FDPSettings = new FDPSettings();
    accessorySettings: AccessorySettings = new AccessorySettings();
    civilworkSettings: CivilworkSettings = new CivilworkSettings();
    ductSettings: DuctSettings = new DuctSettings();
    equipmentSettings: EquipmentSettings = new EquipmentSettings();
    fiberDrop: number;
}

export class EntitySettings {
    entityName: string = "";
    dataSource: string = "automatic";
    existingDesignId?: string;
    importedFile?: string;
    chambers: Chamber[] = new Array<Chamber>();
    odfs: ODF[] = new Array<ODF>();
    poles: Pole[] = new Array<Pole>();
    spliceClosures: SpliceClosure[] = new Array<SpliceClosure>();
    splitters: Splitter[] = new Array<Splitter>();
    pops: PoP[] = new Array<PoP>();
}

export class FDPSettings {
    dataSource: string = "import";
    existingDesignId?: string;
    importedFile?: string;
    fdps: FDP[] = new Array<FDP>();
    aggregateFdps: FDP[] = new Array<FDP>();
}

export class AccessorySettings {
    entityName: string = "";
    dataSource: string = "automatic";
    //existingDesignId?: string;
    //importedFile?: string;
    accessories?: Accessory[] = new Array<Accessory>();
    existingAccessories?: Accessory[] = new Array<Accessory>();
}

export class CivilworkSettings {
    civilworks: Civilwork[] = new Array<Civilwork>();
}

export class DuctSettings {
    ducts: Duct[] = new Array<Duct>();
}

export class EquipmentSettings {
    entityName: string = "";
    dataSource: string = "";
    equipments: string[] = new Array<string>();
}

export class Cabling {
    cableSettings: CableSettings = new CableSettings();
    spliceClosureSettings: SpliceClosureSettings = new SpliceClosureSettings();
    accessorySettings: AccessorySettings = new AccessorySettings();
    ductSettings: DuctSettings = new DuctSettings();
}

export class CableSettings {
    entityName: string = "Cable";
    cableType: string;
    dataSource: string = "";
    cables: Cable[] = new Array<Cable>();
    pole: Pole = new Pole();
    chamber: Chamber = new Chamber();
    chamberInterval: number;
    cableLoop: number;
    loopInterval: number;
    loopAtClosure: number;
    connectionPercent: number = 100;
}

export class SpliceClosureSettings {
    entityName: string = "";
    dataSource: string = "";
    spliceClosures: SpliceClosure[] = new Array<SpliceClosure>();;
    straightJoin: boolean;
    drumLength: number;
    intersectionJoints: boolean;
}

export class CodeId {
    entityType: string;
    maxId: number;
}

export class Costing {
    levelName: string;
    level: number;
    entityType: string;
    code: string;
    description: string;
    units: string;
    quantity: number = 0;
    materialCost: number = 0;
    serviceCost: number = 0;
    currency: string;
}
